feat(lambda-python): introducing LambdaPython

adamelmore opened the pull request:
Higher level API to work with Lambda functions written in Python.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5758d192fdf97e8760f4128f8355a1b5612feb1a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c95fc090-0e36-4f1e-8218-86a7c550848a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5515381977e7415838f13c884bd628622715bcea
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f8b39f17-54fb-4687-bd04-5867e61d7676%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

adamelmore commented:
A few additional notes, if introducing this construct is agreed to be desirable:
- There are chunks of code—mostly utils—that are straight copy/pasted from the Nodejs construct. I considered moving these into `packages/@aws-cdk/aws-lambda/lib/util.ts` but decided to wait for initial feedback.
- I still need to work out some issues with the integration test; I'll add some comments to the test files shortly.
- `requirements.txt` has to be alongside the entry file, meaning you couldn't have multiple Python lambdas (with independent requirements) in the same directory. Should it first look for `stack.my_handler.requirements.txt`, or something along those lines? Should the path to `requirements.txt` be configurable?

adamelmore commented on packages/@aws-cdk/aws-lambda-python/test/integ.function.ts:
```diff
@@ -0,0 +1,30 @@
+/// !cdk-integ cdk-integ-lambda-python
```
This is my first experience with building and testing the CDK locally, and I had a couple of issues.

First, when running `cdk-integ` without this line, I get the following:
```
$ cdk-integ
Synthesizing integ.function.js.
Error: "cdk-integ" can only operate on apps with a single stack.

  If your app has multiple stacks, specify which stack to select by adding this to your test source:

      /// !cdk-integ STACK ...

  Available stacks: Collecting requests==2.23.0   Downloading requests-2.23.0-py2.py3-none-any.whl (58 kB) Collecting pillow==7.2.0   Downloading Pillow-7.2.0-cp36-cp36m-manylinux1_x86_64.whl (2.2 MB) Collecting certifi>=2017.4.17   Downloading certifi-2020.6.20-py2.py3-none-any.whl (156 kB) Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1   Downloading urllib3-1.25.9-py2.py3-none-any.whl (126 kB) Collecting idna<3,>=2.5   Downloading idna-2.10-py2.py3-none-any.whl (58 kB) Collecting chardet<4,>=3.0.2   Downloading chardet-3.0.4-py2.py3-none-any.whl (133 kB) Installing collected packages: certifi, urllib3, idna, chardet, requests, pillow Successfully installed certifi-2020.6.20 chardet-3.0.4 idna-2.10 pillow-7.2.0 requests-2.23.0 urllib3-1.25.9 cdk-integ-lambda-python (wildcards are also supported)
```
I haven't dug into it much, yet, but somehow I've got stdout from my Docker container leaking into the available stacks reported by `cdk-integ` 😅.

adamelmore replied:
The second issue I'm having is that the asset bundle seems to be recreated each time I run the tests, which leads to reported differences in the synth output (asset hashes are changing). This is the cause for the current build failures occurring on this branch.

eladb replied:
Your first issue is a bug (copy @jogold #9186).

I think python should use "SOURCE" hash type because python includes timestamps in the bundled output.

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/function.ts:
```diff
@@ -0,0 +1,103 @@
+   * Path to the (Python) entry file.
+   *
+   * @default - Derived from the name of the defining file and the construct's id.
+   * If the `PythonFunction` is defined in `stack.ts` with `my-handler` as id
```
How would that behave in a CDK python app? Will it use the defining python file name.
Perhaps just make this required to avoid too much magic.

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/function.ts:
```diff
@@ -0,0 +1,103 @@
+   *
+   * @default lambda_handler
+   */
+  readonly handler?: string;
```
I think the convention is just `handler`.

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/function.ts:
```diff
@@ -0,0 +1,103 @@
+   * The runtime environment. Only runtimes of the Python family are
+   * supported.
+   *
+   * @default - `PYTHON_3_7`
```
```suggestion
   * @default lambda.Runtime.PYTHON_3_7
```

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/function.ts:
```diff
@@ -0,0 +1,103 @@
+
+    // Entry and defaults
+    const entry = path.resolve(findEntry(id, props.entry));
+    const handler = props.handler ?? 'lambda_handler';
```
```suggestion
    const handler = props.handler ?? 'handler';
```

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/function.ts:
```diff
@@ -0,0 +1,103 @@
+/**
+ * Finds the name of the file where the `PythonFunction` is defined
+ */
+function findDefiningFile(): string {
```
This will fail for a python CDK app. I would just avoid this magic altogether

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/index.ts:
```diff
@@ -0,0 +1,2 @@
+export * from './function';
+export * from './bundling';
```
Why do we need this exported?

jogold commented on packages/@aws-cdk/aws-lambda-python/lib/bundling.ts:
```diff
@@ -0,0 +1,78 @@
+   *
+   * @default - the closest path containing a .git folder
+   */
+  readonly projectRoot?: string;
```
Why do you need to mount the whole CDK project here? Isn't the assumption here that the Lambda is self contained in a directory with the handler and its requirements?

jogold commented on packages/@aws-cdk/aws-lambda-python/lib/bundling.ts:
```diff
@@ -0,0 +1,78 @@
+    let depsCommand = chain([
+      hasRequirements ? `${installer} install -r requirements.txt -t ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}` : '',
+      `rsync -r . ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}`,
+      `mv ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}/${entryFilename} ${cdk.AssetStaging.BUNDLING_OUTPUT_DIR}/lambda_function.py`,
```
Are you sure you're never going to overwrite an existing file with this `mv`? You can use `entryFilename` instead of the constant `lambda_function` when defining your `handler`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 52e37784e97d18014e870a592ac719a0c99d64bb
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=43dbdc0a-ccd6-484c-bb1b-d6966104547b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

adamelmore commented:
@eladb @jogold I've addressed all of the feedback thus far, and the result is much simpler. `utils.ts` was removed completely which removed the need for any sharing of code between `aws-lambda-python` and `aws-lambda-nodejs`, I believe.

jogold commented on packages/@aws-cdk/aws-lambda-python/lib/function.ts:
```diff
@@ -0,0 +1,75 @@
+/**
+ * Resolves the given entry file.
+ */
+function resolveEntry(entry?: string): string {
+  if (entry) {
+    if (!/\.py$/.test(entry)) {
+      throw new Error('Only Python (.py) entry files are supported.');
+    }
+    if (!fs.existsSync(entry)) {
+      throw new Error(`Cannot find entry file at ${entry}`);
+    }
+    return path.resolve(entry);
+  }
+
+  throw new Error('Cannot find entry file.');
+}
```
The `entry` that you are passing is always a `string` now

```suggestion
function resolveEntry(entry: string): string {
  if (!/\.py$/.test(entry)) {
    throw new Error('Only Python (.py) entry files are supported.');
  }
  if (!fs.existsSync(entry)) {
    throw new Error(`Cannot find entry file at ${entry}`);
  }
  return path.resolve(entry);
}
```

jogold commented on packages/@aws-cdk/aws-lambda-python/package.json:
```diff
@@ -0,0 +1,89 @@
+    "@aws-cdk/assert": "0.0.0",
+    "cdk-build-tools": "0.0.0",
+    "cdk-integ-tools": "0.0.0",
+    "delay": "4.3.0",
```
You can remove this dependency
```suggestion

```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b58601644207cceedb8f9269e24174a42ca11435
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b77dd5b2-b409-43fc-86ba-d564c7b73b63%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f88d0787be29dd135d0ac9e34a01a9cc077e51ec
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e9f49aeb-b3b3-44cf-9b0b-01332d9afdb2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

adamelmore commented:
@jogold I also considered carrying over the notion of `externalModules` here (defaulting to `['botocore', 'boto3']`), but on the python side I've often read that it's best to include these explicitly in your lambda bundle. That sentiment could be due to `requests` being vendored from `botocore`, historically; this [was changed](https://aws.amazon.com/blogs/developer/removing-the-vendored-version-of-requests-from-botocore/) late last year.

Not sure if you have an opinion, but wanted to bring visibility.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: acc28a70fb6283f98e2cbb13ef51b9c51dbc2588
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4dfb6591-eee5-4f40-a441-cfecca3aebba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 078c2757fdd71a039424ba39cd1e996852a3c19c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5079e27b-1731-4a9e-a730-c17a54ee5403%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 078c2757fdd71a039424ba39cd1e996852a3c19c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=150f7889-655d-4073-8944-5efcf3bf3a79%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/bundling.ts:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Bundling
+ */
+export class Bundling {
```
Since this is an internal API, it could just be a function:

```ts
export function bundle(options: BunldingOptions) ...
```

eladb commented on packages/@aws-cdk/aws-lambda-python/lib/function.ts:
```diff
@@ -0,0 +1,71 @@
+  /**
+   * Path to the (Python) entry file.
+   */
+  readonly entry: string;
```
I wonder if it makes more sense that this would be the root directory (where you'd look for `requirements.txt`) and add another (optional) option for the index file? (defaults to `index.py`).

The current situation precludes the option of putting your entry point in a subdirectory...

adamelmore commented:
> I love how simple this is!

Keeps getting simpler! 🎉

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 462cc96babf89533de52b92f3adb12cbf4649bfe
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=241f1a6c-aef7-401c-bf21-8b7f32524eb7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b371905856f37f90ceb68834d9629f232994534e
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5b13723e-be2b-40f4-800c-906400f91f1c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jogold commented:
> @jogold I also considered carrying over the notion of `externalModules` here (defaulting to `['botocore', 'boto3']`), but on the python side I've often read that it's best to include these explicitly in your lambda bundle. That sentiment could be due to `requests` being vendored from `botocore`, historically; this [was changed](https://aws.amazon.com/blogs/developer/removing-the-vendored-version-of-requests-from-botocore/) late last year.
> 
> Not sure if you have an opinion, but wanted to bring visibility.

You can control this by including/omitting `boto3`/`botocore` in your `requirements.txt`, no?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 41be6b3c4a74f9de5ac41f4f8b7cac69844bf911
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=20353c12-953b-4801-9068-a2064369cd82%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 33e6c3333d058e28048ce8ae4032ebbe19c884f6
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9bccf779-6a15-4b55-9460-9f2ed24dd347%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b9f9cd175647bac94a89762d4926953a1993f92d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7757d4d8-dbb9-4d9d-891b-f6a36b367f62%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
